src/game.rs

```diff
@@ -26,26 +30,161 @@ pub enum GameStatus {
+            if self.state.get_piece(&to).is_none() && self.can_double_move(&from) {
+                self.valid_moves.insert(Move { from, to });
+            }
+        }
+    }
+
+    fn can_double_move(&self, from: &Position) -> bool {
+        if let Piece::Pawn(player) = self.state.get_piece(from).unwrap() {
+            return match player {
+                Player::White => from.y == 1,
+                Player::Black => from.y == 6,
+            };
+        }
+        false
+    }
+
+    fn add_pawn_capture_moves(&mut self, from: Position) {
+        for &v in Displacement::get_pawn_capture_vectors(self.state.player) {
+            let to = from + v;
+            if BoardState::is_in_bounds(&to).is_ok() {
+                if let Some(piece) = self.state.get_piece(&to) {
+                    if piece.get_player() != self.state.player {
+                        self.valid_moves.insert(Move::new(from, to));
+                    }
+                }
+                if Some(to) == self.state.en_passant_position {
+                    self.valid_moves.insert(Move::new(from, to));
+                }
+            }
+        }
+    }
+
+    fn add_moves_for_piece(&mut self, from: Position) {
+        if let Some(piece) = self.state.get_piece(&from) {
+            if piece.get_player() == self.state.player {
+                if piece.is_pawn() {
+                    self.add_pawn_advance_moves(from);
+                    self.add_pawn_capture_moves(from);
+                } else {
+                    for &v in piece.get_vectors() {
+                        let mut to = from + v;
+                        while BoardState::is_in_bounds(&to).is_ok() {
+                            if let Some(piece) = self.state.get_piece(&to) {
+                                if piece.get_player() != self.state.player {
+                                    self.valid_moves.insert(Move { from, to });
+                                }
+                                break;
+                            }
+                            self.valid_moves.insert(Move { from, to });
+                            if !self.state.get_piece(&from).unwrap().can_snipe() {
+                                break;
+                            }
+                            to += v;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    pub fn add_moves(&mut self) {
+        self.valid_moves.clear();
+        for y in 0..8 {
+            for x in 0..8 {
+                self.add_moves_for_piece(Position::new(x, y))
+            }
+        }
+        self.add_castle_moves();
+    }
+
+    fn add_castle_moves(&mut self) {
+        let (king_square, kingside, queenside) =
+            CastlingRights::get_castling_info(self.state.player);
+
+        if self.state.castle_rights[kingside as usize]
+            && !(1..=2).any(|i| {
+                self.state
+                    .has_piece(&(king_square + Displacement::RIGHT * i))
```

Comment:
Do you think `game` should have it's own `has_piece`? There's also a place in `chess_widget` where we could use it instead of `get_piece..is_some()`.